service/history/handler.go

```diff
@@ -206,6 +209,22 @@ func (h *Handler) isStopped() bool {
 	return atomic.LoadInt32(&h.status) == common.DaemonStatusStopped
 }
 
+func (h *Handler) HealthCheck(
+	_ context.Context,
+	_ *historyservice.HealthCheckRequest,
+) (_ *historyservice.HealthCheckResponse, retError error) {
+	defer log.CapturePanic(h.logger, &retError)
+	h.startWG.Wait()
+
+	latency := h.persistenceHealthSignal.AverageLatency()
+	errRatio := h.persistenceHealthSignal.ErrorRatio()
+
+	if latency > h.config.HealthPersistenceLatencyFailure() || errRatio > h.config.HealthPersistenceErrorRatio() {
+		return &historyservice.HealthCheckResponse{State: enumsspb.HEALTH_STATE_NOT_SERVING}, nil
```

Comment:
I'm probably missing some context, but what's this healthcheck eventually going to be used for? I'm a little hesitant of a healthcheck that leans heavily on a dependent service's availability to pass, since you can quickly end up in a cascading failure situation if the healthcheck is used to determine things like "did the deployment succeed/service start", or "can this service take requests".

Reply:
This is close to deep health check and check dependencies.